### PR TITLE
Add certificate issue/renewal logic

### PR DESCRIPTION
**Pull request by mattmb**
Issues a certificate via vault if https is enabled for the API endpoint.
If the certificate is expired then we renew it and retry the request.

### CODE REVIEW

File: paasta_tools/api/auth_decorator.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class AuthFutureDecorator:
+
+    def __init__(self, future, cluster_name):
+        self.future = future
+        self.attempts = 0
+        self.cluster_name = cluster_name
+
+    def result(self, *args, **kwargs):
```
Comment (vkhromov) [inline paasta_tools/api/auth_decorator.py]: `def result(self, timeout=None)`?

File: paasta_tools/api/auth_decorator.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        """ Wraps a Future's result method with a timer. """
+        try:
+            result = self.future.result(*args, **kwargs)
```
Comment (vkhromov) [inline paasta_tools/api/auth_decorator.py]: `result = self.future.result(timeout)`?

File: paasta_tools/api/auth_decorator.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            if 'SSL' in e.response.text:
+                if self.attempts >= 3:
+                    raise
+                self.attempts += 1
+                system_paasta_config = load_system_paasta_config()
+                paasta_tools.api.client.renew_issue_cert(
+                    system_paasta_config=system_paasta_config,
+                    cluster=self.cluster_name,
+                )
+                result = self.result(*args, **kwargs)
```
Comment (vkhromov) [inline paasta_tools/api/auth_decorator.py]: It seems that on the first recursion call you'll try to get `result` from the same `self.future` object, right?
If that's true then it isn't clear why there will be a second HTTP request/response with the new cert/key and not just the same exception raised as the result of the previous future execution result.

File: paasta_tools/api/client.py [hunk 4]
```diff
@@ -89,16 +91,41 @@ def __init__(
         system_paasta_config: SystemPaastaConfig,
     ) -> None:
         if scheme == 'https':
+            ca_cert_path = None
             if system_paasta_config.get_enable_client_cert_auth():
+                ecosystem = system_paasta_config.get_vault_cluster_config()[cluster]
                 paasta_dir = os.path.expanduser('~/.paasta/pki')
-                ssl_cert = (f'{paasta_dir}/{cluster}.crt', f'{paasta_dir}/{cluster}.key')
+                if not os.path.isfile(f'{paasta_dir}/{ecosystem}.crt'):
```
Comment (vkhromov) [inline paasta_tools/api/client.py]: probably better to check the existence of all three files: key/crt/ca?

File: paasta_tools/api/client.py [hunk 5]
```diff
@@ -89,16 +91,41 @@ def __init__(
         system_paasta_config: SystemPaastaConfig,
     ) -> None:
         if scheme == 'https':
+            ca_cert_path = None
             if system_paasta_config.get_enable_client_cert_auth():
+                ecosystem = system_paasta_config.get_vault_cluster_config()[cluster]
```
Comment (vkhromov) [inline paasta_tools/api/client.py]: This seems to be assume that we use the same cert/key/ca for all paasta clusters in the ecosystem, right?

File: paasta_tools/api/auth_decorator.py (hunk 3 above)
Comment (mattmb) [reply paasta_tools/api/auth_decorator.py]: Yup, we discussed in real life. I think this is just bad naming, it definitely fires another request with the new cert.

File: paasta_tools/api/client.py (hunk 5 above)
Comment (mattmb) [reply paasta_tools/api/client.py]: Yeah, I think this is ok to start with. I was going back and forth between cluster/ecosystem. But as this is just authn I think ecosystem will be fine and per-cluster permission can be at the authz layer.

**Pull request merged**